lib/ts/recipe/webauthn/api/implementation.ts

```diff
@@ -0,0 +1,1095 @@
+            options: APIOptions;
+            userContext: UserContext;
+            // should also have the email or recoverAccountToken in order to do the preauth checks
+        }): Promise<
+            | {
+                  status: "OK";
+                  session: SessionContainerInterface;
+                  user: User;
+              }
+            | GeneralErrorResponse
+            | {
+                  status: "SIGN_UP_NOT_ALLOWED";
+                  reason: string;
+              }
+            | { status: "EMAIL_ALREADY_EXISTS_ERROR" }
+            | { status: "WRONG_CREDENTIALS_ERROR" }
+            | { status: "INVALID_AUTHENTICATOR_ERROR"; reason: string }
+        > {
+            const errorCodeMap = {
+                SIGN_UP_NOT_ALLOWED:
+                    "Cannot sign up due to security reasons. Please try logging in, use a different login method or contact support. (ERR_CODE_007)",
+                INVALID_AUTHENTICATOR_ERROR: {
+                    // TODO: add more cases
+                },
+                WRONG_CREDENTIALS_ERROR: "The sign up credentials are incorrect. Please use a different authenticator.",
+                LINKING_TO_SESSION_USER_FAILED: {
+                    EMAIL_VERIFICATION_REQUIRED:
+                        "Cannot sign in / up due to security reasons. Please contact support. (ERR_CODE_013)",
+                    RECIPE_USER_ID_ALREADY_LINKED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR:
+                        "Cannot sign in / up due to security reasons. Please contact support. (ERR_CODE_014)",
+                    ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR:
+                        "Cannot sign in / up due to security reasons. Please contact support. (ERR_CODE_015)",
+                    SESSION_USER_ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR:
+                        "Cannot sign in / up due to security reasons. Please contact support. (ERR_CODE_016)",
+                },
+            };
+
+            const generatedOptions = await options.recipeImplementation.getGeneratedOptions({
+                webauthnGeneratedOptionsId,
+                tenantId,
+                userContext,
+            });
+            if (generatedOptions.status !== "OK") {
+                return { status: "WRONG_CREDENTIALS_ERROR" };
+            }
+
+            const email = generatedOptions.email;
+
+            // NOTE: Following checks will likely never throw an error as the
+            // check for type is done in a parent function but they are kept
+            // here to be on the safe side.
+            if (!email) {
+                throw new Error(
+                    "Should never come here since we already check that the email value is a string in validateFormFieldsOrThrowError"
+                );
+            }
+
+            const preAuthCheckRes = await AuthUtils.preAuthChecks({
+                authenticatingAccountInfo: {
```

Comment:
shouldn't this have the credentialId? or is that something we only get later?

Reply:
I don't think so. We are identifying the user by email. And since we are signing up, the credential isn't stored yet.

Reply:
so in no cases would this be relevant to this or other linking related functions? if so then the types may require some cleanup

Reply:
will check if we need to remove webauthn credential ids from the type - it is not used atm